### PR TITLE
fix(core): fix potential data loss when non-WAL table dropped and created with same name

### PR DESCRIPTION
**Pull request by ideoma**
The problem found by fuzz tests. 
When a non-WAL table is re-created with the same name (dropped and created with the same name) the partition purge job can drop a partition of the new table. This is relevant log:

```
2024-05-20T21:52:11.600808Z I i.q.t.f.FuzzDropCreateTableOperation dropping table testWalAddRemoveCommitDropFuzz_nonwal~
2024-05-20T21:52:11.619260Z I i.q.c.p.WriterPool created [table=`testWalAddRemoveCommitDropFuzz_nonwal~`, thread=31]
2024-05-20T21:52:11.619370Z I i.q.c.TableWriter open 'testWalAddRemoveCommitDropFuzz_nonwal'
2024-05-20T21:52:11.622014Z I i.q.c.CairoEngine unlocked [table=`testWalAddRemoveCommitDropFuzz_nonwal`]
2024-05-20T21:52:11.622049Z I i.q.c.p.WriterPool >> [table=`testWalAddRemoveCommitDropFuzz_nonwal~`, thread=31]
2024-05-20T21:52:11.625181Z I i.q.c.TableWriter switched partition [path='/tmp/junit13872316854020729115/dbRoot/testWalAddRemoveCommitDropFuzz_nonwal~/2022-02-25']
2024-05-20T21:52:11.635168Z I i.q.c.O3PartitionPurgeJob purging dropped partition directory [path=/tmp/junit13872316854020729115/dbRoot/testWalAddRemoveCommitDropFuzz_nonwal~/2022-02-24T172921-000001.6]
2024-05-20T21:52:11.635202Z I i.q.c.O3PartitionPurgeJob purging dropped partition directory [path=/tmp/junit13872316854020729115/dbRoot/testWalAddRemoveCommitDropFuzz_nonwal~/2022-02-25]
```

that eventually lead to 

```
2024-05-20T21:52:11.8031636Z io.questdb.cairo.CairoException: [0] Partition '2022-02-25' does not exist in table 'testWalAddRemoveCommitDropFuzz_nonwal' directory. Run [ALTER TABLE testWalAddRemoveCommitDropFuzz_nonwal DROP PARTITION LIST '2022-02-25'] to repair the table or restore the partition directory.
```

The fix is to introduce a table create a lock that `O3PartitionPurgeJob` will hold for a short period when it tries to remove the partition folder. The lock will not allow to create a new table in the partition directory

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 68 / 69 (98.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|57|58|98.28%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|11|11|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 68 / 69 (98.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|57|58|98.28%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|11|11|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 69 / 69 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|58|58|100.00%|

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 1]
```diff
@@ -73,6 +73,7 @@ public class CairoEngine implements Closeable, WriterSource {
     protected final CairoConfiguration configuration;
     private final AtomicLong asyncCommandCorrelationId = new AtomicLong();
     private final CopyContext copyContext;
+    private final ConcurrentHashMap<TableToken> createTableLock = new ConcurrentHashMap<>();
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:76]: We already have a similar map in `TableNameRegistryRW` (`tableNameToTableTokenMap`). So, maybe we could add `boolean lockTableName(TableToken)` method to `TableNameRegistry` and use it `O3PartitionPurgeJob#purgePartition()`?

Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:76]: I considered it but couldn't find a way to make it work.
TableNameRegistry cannot block an existing table from being re-created. Once it's dropped it is available to be created again. To lock the table from creation an additional data structure is needed.

**Review by puzpuzpuz — APPROVED** [review]

**Pull request merged**